pkg/server/auth/jwt.go

```diff
@@ -99,13 +101,22 @@ func extractToken(s string) string {
 	return strings.TrimSpace(parts[1])
 }
 
-func parseJWTToken(ctx context.Context, verifier tokenVerifier, rawIDToken string, cc *ClaimsConfig) (*UserPrincipal, error) {
+func parseJWTToken(ctx context.Context, verifier JWTTokenVerifier, rawIDToken string, oc OIDCConfig) (*UserPrincipal, error) {
 	token, err := verifier.Verify(ctx, rawIDToken)
 	if err != nil {
 		return nil, fmt.Errorf("failed to verify JWT token: %w", err)
 	}
 
-	return cc.PrincipalFromClaims(token)
+	if oc.Audiences != nil {
+		for _, aud := range oc.Audiences {
+			if !stringsContain(token.Audience, aud) {
+				// Do not add the valid audiences to this it would be exposing data
+				return nil, fmt.Errorf("token audience %q is not valid for this request", strings.Join(token.Audience, ", "))
+			}
+		}
+	}
```

Comment:
I'm surprised the verfier.Verify doesn't do this, I guess thats more of an OIDC specific concern than JWT?

Reply:
Yes, it's at the OIDC/Application layer, rather than at the JWT, it's not whether or not the token is valid, it's where it originates that's the issue.